cilium/cmd/kernel_check.go

```diff
@@ -0,0 +1,54 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+	"os"
+	"text/tabwriter"
+
+	"github.com/cilium/cilium/pkg/datapath/linux/probes"
+	"github.com/spf13/cobra"
+)
+
+var kernelCheckCmd = &cobra.Command{
+	Use:   "kernel-check",
+	Short: "Checks whether machine is supporting required kernel capabilities",
+	Run: func(cmd *cobra.Command, args []string) {
+		kernelCheck()
```

Comment:
Oh, the other missing feature here is to get the output as json. If you look at some of the other Cilium CLI commands, I think that this should be fairly easy, you can probably just make a struct which holds both the required / optional maps, then populate these two fields inside the struct, then do similar `if command.OutputJSON() { if err := command.PrintOutput(capabilities); err != nil { ... } return }` logic. You can see some examples in files like `cilium/cmd/bpf_lb_list.go`, just search for `JSON`.

Reply:
This can be nice to be able to build additional tooling on top, because the output is machine-parseable. Again, optional extra feature but I think it'd be nice to have. You can test with `cilium kernel-check -o json`.

Reply:
@joestringer I'm a bit stuck here I have to admit: shall I merge the two maps of required and optional params, thus being possible to call `[...] if err := command.PrintOutput(capabilities); err != nil { [...]` ?

Reply:
Yes, I think Joe's suggestion was to merge the two into a single struct array so it's easier to dump. But you can keep a field to tell if the feature is required, so the JSON produced could be something like:
``` json
[
  {
    "feature": "CONFIG_BPF_SYSCALL",
    "supported": true,
    "required": true
  },
  {
    "feature": "CONFIG_NET_SCH_INGRESS",
    "supported": false,
    "required": true
  },
  ...
]
```

Reply:
@qmonnet to merge the two maps (`requiredParams` and `optionalParams`) into a single map I was thinking to use a merge function like this: 

```
func mergeKernelParamMaps(req map[probes.KernelParam]bool, opt map[probes.KernelParam]bool) map[probes.KernelParam]bool {
	for k, v := range req {
		opt[k] = v
	}
	return opt
}
```
then storing the return value of the function call: 

`var capabilities = mergeKernelParamMaps(requiredParams, optionalParams)`

Next, to convert this now unified map into a `struct` for dumping the `JSON` output I'd use something like that:

```
type KernelParams struct {
	feature string 
	supported bool 
}

for k, v := range capabilities {
	jsondump = append(jsondump, KernelParams{
		feature = k, 
		supported = v,
	})
}
```
Somehow this seems to be a whole lot of boilerplate for the requirement. Does this makes sense at all - or am I completely off track here...?

Reply:
You don't have to create a third map to merge the first two, you can simply “merge” them _into_ your struct. Something like:
``` go
type KernelParamSupport struct {
	feature string
	supported bool
	required bool
}

for f, s := range requiredParams {
	listParams = append(listParams, KernelParamSupport {
		feature = f,
		supported = s,
		required = true,
	})
}
for f, s := range optionalParams {
	listParams = append(listParams, KernelParamSupport {
		feature = f,
		supported = s,
		required = false,
	})
}

...
	json.Marshal(listParams)
...
```

(Not tested.) Note that once you have this list with `feature`, `supported` and `required` fields, you could refactor your existing code to also use it for plain output. I mean replacing this:

``` go
		requiredParams := probeManager.GetRequiredConfig()
		for k, v := range requiredParams {
			fmt.Fprintf(w, "%s\t%t\ttrue\n", k, v)
		}

		optionalParams := probeManager.GetOptionalConfig()
		for k, v := range optionalParams {
			fmt.Fprintf(w, "%s\t%t\tfalse\n", k, v)
		}
```
... with this:
``` go
		listParams := ...
		for _, p := range listParams {
			fmt.Fprintf(w, "%s\t%t\t%t\n", p.feature, p.supported, p.required)
		}
```

Reply:
@qmonnet thanks for the hints. Although the `struct`  being populated, the `JSON` output though empty: 

```
sudo cilium kernel-check -o json
[
  {},
  {},
  {},
  {},
  {},
  {},
  {},
  {},
  {},
  {}
]
```
Any ideas what I'm missing here?

Reply:
The fields in struct `KernelParamSupport` need to be exported (so `Feature`, `Supported`, `Required` instead of `feature`, `supported`, `required`) for this to work. Sorry for misleading you on that with my above snippets.